clients/venice-client/src/main/java/com/linkedin/venice/fastclient/RetriableAvroGenericStoreClient.java

```diff
@@ -266,6 +268,17 @@ private <R extends MultiKeyRequestContext<K, V>, RESPONSE> void retryStreamingMu
 
     Runnable retryTask = () -> { // Look at the remaining keys and setup completion
       if (!pendingKeysFuture.isEmpty()) {
+        Throwable throwable = savedException.get();
```

Comment:
Will it be better to do this when the original request completes rather than when the retry task starts, to free up the scheduler more proactively in case of lot of retries happening?

in retryStreamingMultiKeyRequest:
1. schedule the retry task first: scheduledRetryTask
2. send the original request along and pass in scheduledRetryTask to it
3. when original request completes with 429,  scheduledRetryTask can be cancelled there rather than waiting for retry task to executed and then cancel it.
4. Have the current change as a defensive code just in case the retry task starting while cancelling it.

what do you think?

Reply:
Good call, let me make that change.